luaScripting/src/main/kotlin/com/xingpeds/kross/luaScripting/Lua.kt

```diff
@@ -19,64 +24,62 @@ fun Lua.executeFile(file: File, input: InputStream? = null, output: OutputStream
     this.executeLua(codeAsText, input, output, error)
 }
 
-class GlobalsPool(size: Int) {
-    private val pool = ArrayBlockingQueue<Globals>(size)
 
-    init {
-        repeat(size) {
-            pool.add(createGlobals())
-        }
-    }
+object LuaEngine : Lua {
 
-    private fun createGlobals(): Globals {
-        return Globals().apply {
-            load(BaseLib())
-            load(PackageLib())
-            load(Bit32Lib())
-            load(TableLib())
-            load(StringLib())
-            load(CoroutineLib())
-            LoadState.install(this)
-            LuaC.install(this)
+    val registerFunction = object : TwoArgFunction() {
+        override fun call(nameArg: LuaValue, funcArg: LuaValue): LuaValue {
+            val name = nameArg.checkjstring() ?: throw Exception("function name not supplied")// Get string from nameArg
+            val function = funcArg.checkfunction()
+                ?: throw Exception("lua function not supplied") // Ensure funcArg is a LuaFunction
+            _userFunctions.update {
+                it.toMutableMap().apply { this[name] = function }
+            }
+            return LuaValue.NIL // maybe I should return the function?
         }
-    }
 
-    fun acquire(): Globals {
-        return pool.take()
     }
 
-    fun release(globals: Globals) {
-        pool.put(globals)
+    val global = Globals().apply {
+        load(BaseLib())
+        load(PackageLib())
+        load(Bit32Lib())
+        load(TableLib())
+        load(StringLib())
+        load(CoroutineLib())
+        LoadState.install(this)
+        LuaC.install(this)
+        val krossTable = LuaValue.tableOf()          // Create the `kross` table
+        val apiTable = LuaValue.tableOf()            // Create the `api` table
+        apiTable["register"] = registerFunction      // Add `registerFunction` to `api` table
+        krossTable["api"] = apiTable                 // Add the `api` table to `kross` table
+        this["kross"] = krossTable                   // Add the `kross` table to Globals
     }
-}
-
-object LuaObject : Lua {
-
-    val globalsPool = GlobalsPool(3) // Adjust pool size as needed
 
     override fun executeLua(code: String, input: InputStream?, output: OutputStream?, error: OutputStream?) {
-        val globals = globalsPool.acquire()
 
-        val originalStdout = globals.STDOUT
-        val originalStdin = globals.STDIN
-        val originalStderr = globals.STDERR
+        val originalStdout = global.STDOUT
```

Comment:
changing the pipes on global may or may not work for two lua user funcs in a pipeline